runtime/compiler/runtime/JITServerStatisticsThread.cpp

```diff
@@ -0,0 +1,199 @@
+            if (cpuUtil->isFunctional())
+               {
+               TR_VerboseLog::writeLine(TR_Vlog_JITServer, "CpuLoad %d%% (AvgUsage %d%%) JvmCpu %d%%", cpuUsage, avgCpuUsage, vmCpuUsage);
+               }
+            TR_VerboseLog::vlogRelease();
+            lastStatsTime = crtTime;
+            }
+         }
+      // This thread has been interrupted or StatisticsThreadExitFlag flag was set
+      }
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "Detaching JITServer statistics thread");
+
+   // Will reach here only if the _statisticsThreadExitFlag is set from the stopStatisticsThread().
+   vm->internalVMFunctions->DetachCurrentThread((JavaVM *) vm);
+   statsThreadObj->getStatisticsThreadMonitor()->enter();
+   statsThreadObj->setStatisticsThread(NULL);
+   statsThreadObj->getStatisticsThreadMonitor()->notifyAll();
+   j9thread_exit((J9ThreadMonitor*)statsThreadObj->getStatisticsThreadMonitor()->getVMMonitor());
+
+   return 0;
+   }
+
+void
+JITServerStatisticsThread::startStatisticsThread(J9JavaVM *javaVM)
+   {
+   PORT_ACCESS_FROM_JAVAVM(javaVM);
+
+   _statisticsThreadMonitor = TR::Monitor::create("JITServer-StatisticsThreadMonitor");
+   if (_statisticsThreadMonitor)
+      {
+      const UDATA defaultOSStackSize = javaVM->defaultOSStackSize; //256KB stack size
+      if (javaVM->internalVMFunctions->createThreadWithCategory(&_statisticsOSThread,
+                                                               defaultOSStackSize,
+                                                               J9THREAD_PRIORITY_NORMAL,
+                                                               0,
+                                                               &statisticsThreadProc,
+                                                               javaVM->jitConfig,
+                                                               J9THREAD_CATEGORY_SYSTEM_JIT_THREAD))
+         { // cannot create the statistics thread
+         j9tty_printf(PORTLIB, "Warning: Unable to create JITServer Statistics Thread.\n");
+         TR::Monitor::destroy(_statisticsThreadMonitor);
+         _statisticsThreadMonitor = NULL;
+         }
+      else // must wait here until the thread gets created; otherwise an early shutdown
+         { // does not know whether or not to destroy the thread
+         _statisticsThreadMonitor->enter();
+         while (!getAttachAttempted())
+            _statisticsThreadMonitor->wait();
+         _statisticsThreadMonitor->exit();
+         if (!getStatisticsThread())
+            {
+            j9tty_printf(PORTLIB, "Warning: JITServer Statistics Thread attach failed.\n");
+            }
+         }
+      }
+   else
+      {
+      // JITServer can function without the statistics thread
+      j9tty_printf(PORTLIB, "Warning: Unable to create JITServer Statistics Monitor.\n");
```

Comment:
This is a little confusing. Elsewhere in this patch if we fail to allocate the stats thread object we return an error (and presumably fail startup), but here if we're not able to create the corresponding monitor we let execution continue and the server can start. If we don't need the stats thread shouldn't we allow execution to continue if `statsThreadObj` cannot be allocated?

Reply:
I decided that is better for the caller to decide what it wants to do.
I changed the code so that the caller will test the existence of the statistics thread and fail the JVM if we wanted to start one.